src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,160 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E element, Node<E> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+        } else {
+            if (tail == null) {
+                tail = new Node<>(head, value, null);
+                head.next = tail;
+            } else {
+                tail.next = new Node<>(tail, value, null);
+                tail = tail.next;
+            }
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexAdd(index);
+        Node<T> foundNode = getNode(index);
+        if (foundNode != null) {
+            Node<T> newNode;
+            if (foundNode.prev != null) {
+                newNode = new Node<>(foundNode.prev, value, foundNode);
+                foundNode.prev.next = newNode;
+            } else {
+                newNode = new Node<>(null, value, foundNode);
+            }
+            if (head == foundNode) {
+                head = newNode;
+            }
+            size++;
+        } else {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        return node != null ? node.item : null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        unlink(node);
+        size--;
+        T value = node.item;
+        node = null;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = getIndex(object);
+        if (index != -1) {
+            remove(index);
+            return true;
+        }
         return false;
     }
```

Comment:
U should to move 'size--' to unlink method or add this to 'remove(T object)' method

Reply:
I don't agree with this, because unlink method is just deleting links for nodes and we don't need to change size in that method

Reply:
Yeah, now i see that u use 'remove(int index)' in 'remove(T object)' method, and size decremented right, my mistake

Reply:
But you were also right, if we unlink node, we remove it from our list 👍